// @tsparticles/engine/cjs/Utils/OptionsUtils.js
import { OptionLoader, loadOptions } from "./OptionLoader.js";
import { deepExtend } from "./Utils.js";
import { setRangeValue } from "./MathUtils.js";
export { OptionLoader, loadOptions };
export function loadProperty(obj, key, value) {
    if (value !== undefined) {
        obj[key] = value;
    }
}
export function loadRangeProperty(obj, key, value) {
    if (value !== undefined) {
        obj[key] = setRangeValue(value);
    }
}
export function loadNestedProperty(obj, key, value) {
    if (value !== undefined) {
        obj[key].load(value);
    }
}
export function loadLazyProperty(obj, key, value, factory) {
    if (value !== undefined) {
        const objRecord = obj;
        objRecord[key] ??= factory();
        objRecord[key].load(value);
    }
}
export function loadExtendProperty(obj, key, value) {
    if (value !== undefined) {
        obj[key] = deepExtend(obj[key] ?? {}, value);
    }
}
export function loadOptionProperty(obj, key, optionClass, ...sources) {
    const objRecord = obj;
    objRecord[key] ??= new optionClass();
    const target = objRecord[key];
    for (const source of sources) {
        target.load(source?.[key]);
    }
}
