// @tsparticles/shape-image/tsparticles.shape.image.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Shape v4.3.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.shapes = global.__tsParticlesInternals.shapes || {}, global.__tsParticlesInternals.shapes.image = global.__tsParticlesInternals.shapes.image || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    const shapeTypes = ["image", "images"];
    const stringStart = 0, defaultOpacity = 1;
    const currentColorRegex = /(#(?:[0-9a-f]{2}){2,4}|(#[0-9a-f]{3})|(rgb|hsl)a?\((-?\d+%?[,\s]+){2,3}\s*[\d.]+%?\))|currentcolor/gi;
    function replaceColorSvg(imageShape, color, opacity, hdr = false) {
        const { svgData } = imageShape;
        if (!svgData) {
            return "";
        }
        const colorStyle = engine.getStyleFromHsl(color, hdr, opacity);
        if (svgData.includes("fill")) {
            return svgData.replaceAll(currentColorRegex, () => colorStyle);
        }
        const preFillIndex = svgData.indexOf(">");
        return `${svgData.substring(stringStart, preFillIndex)} fill="${colorStyle}"${svgData.substring(preFillIndex)}`;
    }
    async function loadImage(image) {
        return new Promise(resolve => {
            image.loading = true;
            const img = new Image();
            image.element = img;
            img.addEventListener("load", () => {
                image.loading = false;
                resolve();
            });
            img.addEventListener("error", () => {
                image.element = undefined;
                image.error = true;
                image.loading = false;
                engine.getLogger().error(`Error loading image: ${image.source}`);
                resolve();
            });
            img.src = image.source;
        });
    }
    async function downloadSvgImage(image) {
        if (image.type !== "svg") {
            await loadImage(image);
            return;
        }
        image.loading = true;
        const response = await fetch(image.source);
        if (response.ok) {
            image.svgData = await response.text();
        }
        else {
            engine.getLogger().error("Image not found");
            image.error = true;
        }
        image.loading = false;
    }
    function replaceImageColor(image, imageData, color, particle, hdr = false) {
        const svgColoredData = replaceColorSvg(image, color, particle.opacity?.value ?? defaultOpacity, hdr), imageRes = {
            color,
            data: {
                ...image,
                svgData: svgColoredData,
            },
            loaded: false,
            ratio: imageData.width / imageData.height,
            replaceColor: imageData.replaceColor,
            source: imageData.src,
        };
        return new Promise(resolve => {
            const svg = new Blob([svgColoredData], { type: "image/svg+xml" }), url = URL.createObjectURL(svg), img = new Image();
            img.addEventListener("load", () => {
                imageRes.loaded = true;
                imageRes.element = img;
                resolve(imageRes);
                URL.revokeObjectURL(url);
            });
            const errorHandler = async () => {
                URL.revokeObjectURL(url);
                const img2 = {
                    ...image,
                    error: false,
                    loading: true,
                };
                await loadImage(img2);
                imageRes.loaded = true;
                imageRes.element = img2.element;
                resolve(imageRes);
            };
            img.addEventListener("error", () => void errorHandler());
            img.src = url;
        });
    }

    const sides = 12;
    class ImageDrawer {
        #container;
        #engine;
        constructor(engine, container) {
            this.#engine = engine;
            this.#container = container;
        }
        draw(data) {
            const { context, radius, particle, opacity } = data, image = particle.image, element = image?.element;
            if (!image) {
                return;
            }
            context.globalAlpha = opacity;
            if (element) {
                const ratio = image.ratio, pos = {
                    x: -radius,
                    y: -radius,
                }, diameter = radius * engine.double;
                context.drawImage(element, pos.x, pos.y, diameter, diameter / ratio);
            }
            context.globalAlpha = engine.defaultAlpha;
        }
        getSidesCount() {
            return sides;
        }
        async init(container) {
            const options = container.actualOptions;
            if (!options.preload || !this.#engine.loadImage) {
                return;
            }
            const promises = [];
            for (const imageData of options.preload) {
                promises.push(this.#engine.loadImage(container, imageData));
            }
            await Promise.all(promises);
        }
        loadShape(particle) {
            const container = this.#container;
            if (!particle.shape || !shapeTypes.includes(particle.shape)) {
                return;
            }
            const imageData = particle.shapeData;
            if (!imageData) {
                return;
            }
            const images = this.#engine.getImages?.(container), image = images?.find((t) => t.name === imageData.name || t.source === imageData.src);
            if (image) {
                return;
            }
            void this.#loadImageShape(container, imageData).then(() => {
                this.loadShape(particle);
            });
        }
        particleInit(container, particle) {
            if (particle.shape !== "image" && particle.shape !== "images") {
                return;
            }
            const images = this.#engine.getImages?.(container), imageData = particle.shapeData;
            if (!imageData) {
                return;
            }
            const color = particle.getFillColor(), image = images?.find((t) => t.name === imageData.name || t.source === imageData.src);
            if (!image) {
                return;
            }
            const replaceColor = imageData.replaceColor;
            if (image.loading) {
                setTimeout(() => {
                    this.particleInit(container, particle);
                });
                return;
            }
            void (async () => {
                let imageRes;
                if (image.svgData && color) {
                    imageRes = await replaceImageColor(image, imageData, color, particle, container.hdr);
                }
                else {
                    imageRes = {
                        color,
                        data: image,
                        element: image.element,
                        loaded: true,
                        ratio: imageData.width && imageData.height ? imageData.width / imageData.height : (image.ratio ?? engine.defaultRatio),
                        replaceColor: replaceColor,
                        source: imageData.src,
                    };
                }
                if (!imageRes.ratio) {
                    imageRes.ratio = 1;
                }
                const close = imageData.close ?? particle.shapeClose, imageShape = {
                    image: imageRes,
                    close,
                };
                particle.image = imageShape.image;
                particle.shapeClose = imageShape.close;
            })();
        }
        #loadImageShape = async (container, imageShape) => {
            if (!this.#engine.loadImage) {
                throw new Error(`Image shape not initialized`);
            }
            await this.#engine.loadImage(container, {
                name: imageShape.name,
                replaceColor: imageShape.replaceColor,
                src: imageShape.src,
            });
        };
    }

    class Preload {
        height;
        name;
        replaceColor;
        src = "";
        width;
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            engine.loadProperty(this, "height", data.height);
            engine.loadProperty(this, "name", data.name);
            engine.loadProperty(this, "replaceColor", data.replaceColor);
            engine.loadProperty(this, "src", data.src);
            engine.loadProperty(this, "width", data.width);
        }
    }

    class ImagePreloaderPlugin {
        id = "image-preloader";
        #engine;
        constructor(engine) {
            this.#engine = engine;
        }
        async getPlugin(container) {
            const { ImagePreloaderInstance } = await Promise.resolve().then(function () { return ImagePreloaderInstance$1; });
            return new ImagePreloaderInstance(this.#engine, container);
        }
        loadOptions(_container, options, source) {
            if (!source?.preload) {
                return;
            }
            options.preload ??= [];
            const preloadOptions = options.preload;
            for (const item of source.preload) {
                const existing = preloadOptions.find(t => t.name === item.name || t.src === item.src);
                if (existing) {
                    existing.load(item);
                }
                else {
                    const preload = new Preload();
                    preload.load(item);
                    preloadOptions.push(preload);
                }
            }
        }
        needsPlugin() {
            return true;
        }
    }

    const extLength = 3;
    function addLoadImageToEngine(engine) {
        engine.getImages ??= (container) => {
            engine.images ??= new Map();
            let images = engine.images.get(container);
            if (!images) {
                images = [];
                engine.images.set(container, images);
            }
            return images;
        };
        engine.loadImage ??= async (container, data) => {
            if (!engine.getImages) {
                throw new Error("No images collection found");
            }
            if (!data.name && !data.src) {
                throw new Error("No image source provided");
            }
            engine.images ??= new Map();
            const containerImages = engine.getImages(container);
            if (containerImages.some((t) => t.name === data.name || t.source === data.src)) {
                return;
            }
            try {
                const image = {
                    name: data.name ?? data.src,
                    source: data.src,
                    type: data.src.substring(data.src.length - extLength),
                    error: false,
                    loading: true,
                    replaceColor: data.replaceColor,
                    ratio: data.width && data.height ? data.width / data.height : undefined,
                };
                containerImages.push(image);
                engine.images.set(container, containerImages);
                let imageFunc;
                if (data.replaceColor) {
                    imageFunc = downloadSvgImage;
                }
                else {
                    imageFunc = loadImage;
                }
                await imageFunc(image);
            }
            catch {
                throw new Error(`${data.name ?? data.src} not found`);
            }
        };
    }
    async function loadImageShape(engine) {
        engine.checkVersion("4.3.1");
        await engine.pluginManager.register(e => {
            addLoadImageToEngine(e);
            e.pluginManager.addPlugin(new ImagePreloaderPlugin(e));
            e.pluginManager.addShape(shapeTypes, container => Promise.resolve(new ImageDrawer(e, container)));
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadImageShape = loadImageShape;

    class ImagePreloaderInstance {
        #container;
        #engine;
        constructor(engine, container) {
            this.#engine = engine;
            this.#container = container;
        }
        destroy() {
            this.#engine.images?.delete(this.#container);
        }
    }

    var ImagePreloaderInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        ImagePreloaderInstance: ImagePreloaderInstance
    });

    exports.loadImageShape = loadImageShape;

}));
Object.assign(globalThis.window || globalThis, { loadImageShape: (globalThis.__tsParticlesInternals.shapes.image || {}).loadImageShape });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
