// @tsparticles/interaction-external-bubble/tsparticles.interaction.external.bubble.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.3.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalBubble = global.__tsParticlesInternals.interactions.externalBubble || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    class BubbleBase {
        color;
        distance = 200;
        duration = 0.4;
        mix = false;
        opacity;
        size;
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            engine.loadProperty(this, "distance", data.distance);
            engine.loadProperty(this, "duration", data.duration);
            engine.loadProperty(this, "mix", data.mix);
            engine.loadRangeProperty(this, "opacity", data.opacity);
            if (data.color !== undefined) {
                const sourceColor = engine.isArray(this.color) ? undefined : this.color;
                this.color = engine.executeOnSingleOrMultiple(data.color, color => {
                    return engine.OptionsColor.create(sourceColor, color);
                });
            }
            engine.loadProperty(this, "size", data.size);
        }
    }

    class BubbleDiv extends BubbleBase {
        selectors = [];
        load(data) {
            super.load(data);
            if (engine.isNull(data)) {
                return;
            }
            engine.loadProperty(this, "selectors", data.selectors);
        }
    }

    class Bubble extends BubbleBase {
        divs;
        load(data) {
            super.load(data);
            if (engine.isNull(data)) {
                return;
            }
            this.divs = engine.executeOnSingleOrMultiple(data.divs, div => {
                const tmp = new BubbleDiv();
                tmp.load(div);
                return tmp;
            });
        }
    }

    class BubbleModifier {
        div;
        enabled = false;
        fillColor;
        finalColor;
        id = "bubble";
        inRange = false;
        opacity;
        priority = 100;
        radius;
        strokeColor;
    }

    var ProcessBubbleType;
    (function (ProcessBubbleType) {
        ProcessBubbleType["color"] = "color";
        ProcessBubbleType["opacity"] = "opacity";
        ProcessBubbleType["size"] = "size";
    })(ProcessBubbleType || (ProcessBubbleType = {}));

    function calculateBubbleValue(particleValue, modeValue, optionsValue, ratio) {
        if (modeValue >= optionsValue) {
            const value = particleValue + (modeValue - optionsValue) * ratio;
            return engine.clamp(value, particleValue, modeValue);
        }
        else if (modeValue < optionsValue) {
            const value = particleValue - (optionsValue - modeValue) * ratio;
            return engine.clamp(value, modeValue, particleValue);
        }
        return undefined;
    }

    const bubbleMode = "bubble", minDistance = 0, defaultClickTime = 0, defaultOpacity = 1, ratioOffset = 1, defaultBubbleValue = 0, minRatio = 0, defaultRatio = 1;
    class Bubbler extends pluginInteractivity.ExternalInteractorBase {
        handleClickMode;
        #maxDistance;
        #modifiers = new WeakMap();
        #pluginManager;
        constructor(pluginManager, container) {
            super(container);
            this.#pluginManager = pluginManager;
            this.#maxDistance = 0;
            container.bubble ??= {};
            this.handleClickMode = (mode) => {
                if (mode !== bubbleMode) {
                    return;
                }
                container.bubble ??= {};
                container.bubble.clicking = true;
            };
        }
        get maxDistance() {
            return this.#maxDistance;
        }
        clear(particle, _delta, force) {
            const mod = this.#modifiers.get(particle);
            if (mod?.inRange && !force) {
                return;
            }
            particle.removeModifier(bubbleMode);
            this.#modifiers.delete(particle);
        }
        getOrCreateModifier(particle) {
            let mod = this.#modifiers.get(particle);
            if (!mod) {
                mod = new BubbleModifier();
                this.#modifiers.set(particle, mod);
                particle.addModifier(mod);
            }
            return mod;
        }
        init() {
            const container = this.container, bubble = container.actualOptions.interactivity?.modes.bubble;
            if (!bubble) {
                return;
            }
            this.#maxDistance = bubble.distance;
            container.retina.bubbleModeDistance = bubble.distance * container.retina.pixelRatio;
            if (bubble.size !== undefined) {
                container.retina.bubbleModeSize = bubble.size * container.retina.pixelRatio;
            }
        }
        interact(interactivityData, delta) {
            const options = this.container.actualOptions, events = options.interactivity?.events;
            if (!events) {
                return;
            }
            const onHover = events.onHover, onClick = events.onClick, hoverEnabled = onHover.enable, hoverMode = onHover.mode, clickEnabled = onClick.enable, clickMode = onClick.mode, divs = events.onDiv;
            if (hoverEnabled && engine.isInArray(bubbleMode, hoverMode)) {
                this.#hoverBubble(interactivityData);
            }
            else if (clickEnabled && engine.isInArray(bubbleMode, clickMode)) {
                this.#clickBubble(interactivityData);
            }
            else {
                pluginInteractivity.divModeExecute(bubbleMode, divs, (selector, div) => {
                    this.#singleSelectorHover(interactivityData, delta, selector, div);
                });
            }
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
            if (!events) {
                return false;
            }
            const { onClick, onDiv, onHover } = events, divBubble = pluginInteractivity.isDivModeEnabled(bubbleMode, onDiv);
            if (!(divBubble || (onHover.enable && !!mouse.position) || (onClick.enable && mouse.clickPosition))) {
                return false;
            }
            return engine.isInArray(bubbleMode, onHover.mode) || engine.isInArray(bubbleMode, onClick.mode) || divBubble;
        }
        loadModeOptions(options, ...sources) {
            engine.loadOptionProperty(options, "bubble", Bubble, ...sources);
        }
        reset(_interactivityData, particle) {
            const mod = this.#modifiers.get(particle);
            if (mod) {
                mod.enabled = false;
                mod.inRange = false;
            }
        }
        #clickBubble(interactivityData) {
            const container = this.container, options = container.actualOptions, mouseClickPos = interactivityData.mouse.clickPosition, bubbleOptions = options.interactivity?.modes.bubble;
            if (!bubbleOptions || !mouseClickPos) {
                return;
            }
            container.bubble ??= {};
            const distance = container.retina.bubbleModeDistance;
            if (!distance || distance < minDistance) {
                return;
            }
            const query = container.particles.grid.queryCircle(mouseClickPos, distance, p => this.isEnabled(interactivityData, p)), { bubble } = container;
            for (const particle of query) {
                if (!bubble.clicking) {
                    continue;
                }
                const mod = this.getOrCreateModifier(particle);
                mod.enabled = !bubble.durationEnd;
                mod.inRange = !bubble.durationEnd;
                const pos = particle.getPosition(), distMouse = engine.getDistance(pos, mouseClickPos), timeSpent = (performance.now() - (interactivityData.mouse.clickTime ?? defaultClickTime)) / engine.millisecondsToSeconds;
                if (timeSpent > bubbleOptions.duration) {
                    bubble.durationEnd = true;
                }
                if (timeSpent > bubbleOptions.duration * engine.double) {
                    bubble.clicking = false;
                    bubble.durationEnd = false;
                }
                const sizeData = {
                    bubbleObj: {
                        optValue: container.retina.bubbleModeSize,
                        value: mod.radius,
                    },
                    particlesObj: {
                        optValue: particle.size.max,
                        value: particle.size.value,
                    },
                    type: ProcessBubbleType.size,
                };
                this.#process(particle, distMouse, timeSpent, sizeData);
                const opacityData = {
                    bubbleObj: {
                        optValue: bubbleOptions.opacity,
                        value: mod.opacity,
                    },
                    particlesObj: {
                        optValue: particle.opacity?.max ?? defaultOpacity,
                        value: particle.opacity?.value ?? defaultOpacity,
                    },
                    type: ProcessBubbleType.opacity,
                };
                this.#process(particle, distMouse, timeSpent, opacityData);
                if (!bubble.durationEnd && distMouse <= distance) {
                    this.#hoverBubbleColor(particle, distMouse);
                }
                else {
                    mod.fillColor = undefined;
                    mod.strokeColor = undefined;
                }
            }
        }
        #hoverBubble(interactivityData) {
            const container = this.container, mousePos = interactivityData.mouse.position, distance = container.retina.bubbleModeDistance;
            if (!distance || distance < minDistance || !mousePos) {
                return;
            }
            const query = container.particles.grid.queryCircle(mousePos, distance, p => this.isEnabled(interactivityData, p));
            for (const particle of query) {
                const mod = this.getOrCreateModifier(particle);
                mod.enabled = true;
                mod.inRange = true;
                const pos = particle.getPosition(), pointDistance = engine.getDistance(pos, mousePos), ratio = ratioOffset - pointDistance / distance;
                if (pointDistance <= distance) {
                    if (ratio >= minRatio && interactivityData.status === pluginInteractivity.mouseMoveEvent) {
                        this.#hoverBubbleSize(particle, ratio);
                        this.#hoverBubbleOpacity(particle, ratio);
                        this.#hoverBubbleColor(particle, ratio);
                    }
                }
                else {
                    this.reset(interactivityData, particle);
                }
                if (interactivityData.status === pluginInteractivity.mouseLeaveEvent) {
                    this.reset(interactivityData, particle);
                }
            }
        }
        #hoverBubbleColor(particle, ratio, divBubble) {
            const options = this.container.actualOptions, bubbleOptions = divBubble ?? options.interactivity?.modes.bubble, mod = this.getOrCreateModifier(particle);
            if (!bubbleOptions) {
                return;
            }
            if (!mod.finalColor) {
                const modeColor = bubbleOptions.color;
                if (!modeColor) {
                    return;
                }
                const bubbleColor = engine.itemFromSingleOrMultiple(modeColor);
                mod.finalColor = engine.rangeColorToHsl(this.#pluginManager, bubbleColor);
            }
            if (!mod.finalColor) {
                return;
            }
            if (bubbleOptions.mix) {
                mod.fillColor = undefined;
                mod.strokeColor = undefined;
                const pColor = particle.getFillColor();
                if (pColor) {
                    const mixedColor = engine.rgbToHsl(engine.colorMix(pColor, mod.finalColor, ratioOffset - ratio, ratio));
                    mod.fillColor = mixedColor;
                    mod.strokeColor = mixedColor;
                }
                else {
                    mod.fillColor = mod.finalColor;
                    mod.strokeColor = mod.finalColor;
                }
            }
            else {
                mod.fillColor = mod.finalColor;
                mod.strokeColor = mod.finalColor;
            }
        }
        #hoverBubbleOpacity(particle, ratio, divBubble) {
            const container = this.container, options = container.actualOptions, modeOpacity = divBubble?.opacity ?? options.interactivity?.modes.bubble?.opacity;
            if (!modeOpacity) {
                return;
            }
            const pOpacity = particle.opacity?.value ?? defaultOpacity, opacity = calculateBubbleValue(pOpacity, modeOpacity, particle.opacity?.max ?? defaultOpacity, ratio);
            if (opacity !== undefined) {
                const mod = this.getOrCreateModifier(particle);
                mod.opacity = opacity;
            }
        }
        #hoverBubbleSize(particle, ratio, divBubble) {
            const container = this.container, modeSize = divBubble?.size ? divBubble.size * container.retina.pixelRatio : container.retina.bubbleModeSize;
            if (modeSize === undefined) {
                return;
            }
            const pSize = particle.size.value, size = calculateBubbleValue(pSize, modeSize, particle.size.max, ratio);
            if (size !== undefined) {
                const mod = this.getOrCreateModifier(particle);
                mod.radius = size;
            }
        }
        #process(particle, distMouse, timeSpent, data) {
            const container = this.container, bubbleParam = data.bubbleObj.optValue, options = container.actualOptions, bubbleOptions = options.interactivity?.modes.bubble;
            if (!bubbleOptions || bubbleParam === undefined) {
                return;
            }
            const bubbleDuration = bubbleOptions.duration, bubbleDistance = container.retina.bubbleModeDistance, particlesParam = data.particlesObj.optValue, pObjBubble = data.bubbleObj.value, pObj = data.particlesObj.value ?? defaultBubbleValue, type = data.type;
            if (!bubbleDistance || bubbleDistance < minDistance || bubbleParam === particlesParam) {
                return;
            }
            container.bubble ??= {};
            const mod = this.getOrCreateModifier(particle);
            if (container.bubble.durationEnd) {
                if (pObjBubble) {
                    if (type === ProcessBubbleType.size) {
                        mod.radius = undefined;
                    }
                    if (type === ProcessBubbleType.opacity) {
                        mod.opacity = undefined;
                    }
                }
            }
            else {
                if (distMouse <= bubbleDistance) {
                    const obj = pObjBubble ?? pObj;
                    if (obj !== bubbleParam) {
                        const value = pObj - (timeSpent * (pObj - bubbleParam)) / bubbleDuration;
                        if (type === ProcessBubbleType.size) {
                            mod.radius = value;
                        }
                        if (type === ProcessBubbleType.opacity) {
                            mod.opacity = value;
                        }
                    }
                }
                else {
                    if (type === ProcessBubbleType.size) {
                        mod.radius = undefined;
                    }
                    if (type === ProcessBubbleType.opacity) {
                        mod.opacity = undefined;
                    }
                }
            }
        }
        #singleSelectorHover(interactivityData, delta, selector, div) {
            const container = this.container, selectors = engine.safeDocument().querySelectorAll(selector), bubble = container.actualOptions.interactivity?.modes.bubble;
            if (!bubble || !selectors.length) {
                return;
            }
            selectors.forEach(item => {
                const elem = item, pxRatio = container.retina.pixelRatio, pos = {
                    x: (elem.offsetLeft + elem.offsetWidth * engine.half) * pxRatio,
                    y: (elem.offsetTop + elem.offsetHeight * engine.half) * pxRatio,
                }, repulseRadius = elem.offsetWidth * engine.half * pxRatio, area = div.type === pluginInteractivity.DivType.circle
                    ? new engine.Circle(pos.x, pos.y, repulseRadius)
                    : new engine.Rectangle(elem.offsetLeft * pxRatio, elem.offsetTop * pxRatio, elem.offsetWidth * pxRatio, elem.offsetHeight * pxRatio), query = container.particles.grid.query(area, p => this.isEnabled(interactivityData, p));
                for (const particle of query) {
                    if (!area.contains(particle.getPosition())) {
                        continue;
                    }
                    const mod = this.getOrCreateModifier(particle);
                    mod.enabled = true;
                    mod.inRange = true;
                    const divs = bubble.divs, divBubble = pluginInteractivity.divMode(divs, elem);
                    if (!mod.div || mod.div !== elem) {
                        this.clear(particle, delta, true);
                        mod.div = elem;
                    }
                    this.#hoverBubbleSize(particle, defaultRatio, divBubble);
                    this.#hoverBubbleOpacity(particle, defaultRatio, divBubble);
                    this.#hoverBubbleColor(particle, defaultRatio, divBubble);
                }
            });
        }
    }

    async function loadExternalBubbleInteraction(engine) {
        engine.checkVersion("4.3.1");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalBubble", container => {
                return Promise.resolve(new Bubbler(e.pluginManager, container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalBubbleInteraction = loadExternalBubbleInteraction;

    exports.Bubble = Bubble;
    exports.BubbleBase = BubbleBase;
    exports.BubbleDiv = BubbleDiv;
    exports.BubbleModifier = BubbleModifier;
    exports.loadExternalBubbleInteraction = loadExternalBubbleInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalBubbleInteraction: (globalThis.__tsParticlesInternals.interactions.externalBubble || {}).loadExternalBubbleInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
