// @tsparticles/interaction-external-slow/cjs/Slower.js
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { getDistance, isInArray, loadOptionProperty, } from "@tsparticles/engine";
import { Slow } from "./Options/Classes/Slow.js";
import { SlowModifier } from "./SlowModifier.js";
const slowMode = "slow", minRadius = 0;
export class Slower extends ExternalInteractorBase {
    #maxDistance;
    #modifiers = new WeakMap();
    constructor(container) {
        super(container);
        this.#maxDistance = 0;
    }
    get maxDistance() {
        return this.#maxDistance;
    }
    clear(particle, _delta, force) {
        const mod = this.#modifiers.get(particle);
        if (mod?.enabled && !force) {
            return;
        }
        particle.removeModifier(slowMode);
        this.#modifiers.delete(particle);
    }
    getOrCreateModifier(particle) {
        let mod = this.#modifiers.get(particle);
        if (!mod) {
            mod = new SlowModifier();
            this.#modifiers.set(particle, mod);
            particle.addModifier(mod);
        }
        return mod;
    }
    init() {
        const container = this.container, slow = container.actualOptions.interactivity?.modes.slow;
        if (!slow) {
            return;
        }
        this.#maxDistance = slow.radius;
        container.retina.slowModeRadius = slow.radius * container.retina.pixelRatio;
    }
    interact() {
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
        return !!events?.onHover.enable && !!mouse.position && isInArray(slowMode, events.onHover.mode);
    }
    loadModeOptions(options, ...sources) {
        loadOptionProperty(options, "slow", Slow, ...sources);
    }
    reset(interactivityData, particle) {
        const mod = this.#modifiers.get(particle);
        if (mod) {
            mod.enabled = false;
        }
        const container = this.container, options = container.actualOptions, mousePos = interactivityData.mouse.position, radius = container.retina.slowModeRadius, slowOptions = options.interactivity?.modes.slow;
        if (!slowOptions || !radius || radius < minRadius || !mousePos) {
            return;
        }
        const particlePos = particle.getPosition(), dist = getDistance(mousePos, particlePos), proximityFactor = dist / radius, slowFactor = slowOptions.factor;
        if (dist > radius) {
            return;
        }
        const activeMod = this.getOrCreateModifier(particle);
        activeMod.enabled = true;
        activeMod.speedFactor = proximityFactor / slowFactor;
    }
}
