// @tsparticles/updater-roll/tsparticles.updater.roll.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Updater v4.3.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.updaters = global.__tsParticlesInternals.updaters || {}, global.__tsParticlesInternals.updaters.roll = global.__tsParticlesInternals.updaters.roll || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    var RollMode;
    (function (RollMode) {
        RollMode["both"] = "both";
        RollMode["horizontal"] = "horizontal";
        RollMode["vertical"] = "vertical";
    })(RollMode || (RollMode = {}));

    const maxAngle = 360;
    function initParticle(pluginManager, particle) {
        const rollOpt = particle.options.roll;
        if (!rollOpt?.enable) {
            particle.roll = {
                enable: false,
                horizontal: false,
                vertical: false,
                angle: 0,
                speed: 0,
            };
            return;
        }
        particle.roll = {
            enable: rollOpt.enable,
            horizontal: rollOpt.mode === RollMode.horizontal || rollOpt.mode === RollMode.both,
            vertical: rollOpt.mode === RollMode.vertical || rollOpt.mode === RollMode.both,
            angle: engine.getRandom() * engine.doublePI,
            speed: engine.getRangeValue(rollOpt.speed) / maxAngle,
        };
        if (rollOpt.backColor) {
            particle.backColor = engine.rangeColorToHsl(pluginManager, rollOpt.backColor);
        }
        else if (rollOpt.darken.enable && rollOpt.enlighten.enable) {
            const alterType = engine.getRandom() >= engine.half ? engine.AlterType.darken : engine.AlterType.enlighten;
            particle.roll.alter = {
                type: alterType,
                value: engine.getRangeValue(alterType === engine.AlterType.darken ? rollOpt.darken.value : rollOpt.enlighten.value),
            };
        }
        else if (rollOpt.darken.enable) {
            particle.roll.alter = {
                type: engine.AlterType.darken,
                value: engine.getRangeValue(rollOpt.darken.value),
            };
        }
        else if (rollOpt.enlighten.enable) {
            particle.roll.alter = {
                type: engine.AlterType.enlighten,
                value: engine.getRangeValue(rollOpt.enlighten.value),
            };
        }
    }
    function updateRoll(particle, delta) {
        const roll = particle.options.roll, data = particle.roll;
        if (!data || !roll?.enable) {
            return;
        }
        const speed = data.speed * delta.factor, max = engine.doublePI;
        data.angle += speed;
        if (data.angle > max) {
            data.angle -= max;
        }
    }

    class RollLight {
        enable = false;
        value = 0;
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            engine.loadProperty(this, "enable", data.enable);
            engine.loadRangeProperty(this, "value", data.value);
        }
    }

    class Roll {
        backColor;
        darken = new RollLight();
        enable = false;
        enlighten = new RollLight();
        mode = RollMode.vertical;
        speed = 25;
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.backColor !== undefined) {
                this.backColor = engine.OptionsColor.create(this.backColor, data.backColor);
            }
            this.darken.load(data.darken);
            engine.loadProperty(this, "enable", data.enable);
            this.enlighten.load(data.enlighten);
            engine.loadProperty(this, "mode", data.mode);
            engine.loadRangeProperty(this, "speed", data.speed);
        }
    }

    class RollUpdater {
        #pluginManager;
        constructor(pluginManager) {
            this.#pluginManager = pluginManager;
        }
        getTransformValues(particle) {
            const roll = particle.roll?.enable && particle.roll, rollHorizontal = roll && roll.horizontal, rollVertical = roll && roll.vertical;
            return {
                a: rollHorizontal ? Math.cos(roll.angle) : undefined,
                d: rollVertical ? Math.sin(roll.angle) : undefined,
            };
        }
        init(particle) {
            initParticle(this.#pluginManager, particle);
        }
        isEnabled(particle) {
            const roll = particle.options.roll;
            return !particle.destroyed && !particle.spawning && !!roll?.enable;
        }
        loadOptions(options, ...sources) {
            engine.loadOptionProperty(options, "roll", Roll, ...sources);
        }
        update(particle, delta) {
            if (!this.isEnabled(particle)) {
                return;
            }
            updateRoll(particle, delta);
        }
    }

    async function loadRollUpdater(engine) {
        engine.checkVersion("4.3.1");
        await engine.pluginManager.register(e => {
            e.pluginManager.addParticleUpdater("roll", () => {
                return Promise.resolve(new RollUpdater(e.pluginManager));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadRollUpdater = loadRollUpdater;

    exports.loadRollUpdater = loadRollUpdater;

}));
Object.assign(globalThis.window || globalThis, { loadRollUpdater: (globalThis.__tsParticlesInternals.updaters.roll || {}).loadRollUpdater });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
