// @tsparticles/plugin-absorbers/browser/Options/Classes/Absorber.js
import { OptionsColor, isNull, loadProperty, setRangeValue, } from "@tsparticles/engine";
import { AbsorberLife } from "./AbsorberLife.js";
import { AbsorberSize } from "./AbsorberSize.js";
export class Absorber {
    color;
    destroy;
    draggable;
    life;
    name;
    opacity;
    orbits;
    position;
    size;
    constructor() {
        this.color = new OptionsColor();
        this.color.value = "#000000";
        this.draggable = false;
        this.opacity = 1;
        this.destroy = true;
        this.orbits = false;
        this.life = new AbsorberLife();
        this.size = new AbsorberSize();
    }
    load(data) {
        if (isNull(data)) {
            return;
        }
        if (data.color !== undefined) {
            this.color = OptionsColor.create(this.color, data.color);
        }
        loadProperty(this, "draggable", data.draggable);
        if (data.life !== undefined) {
            this.life.load(data.life);
        }
        this.name = data.name;
        loadProperty(this, "opacity", data.opacity);
        if (data.position !== undefined) {
            this.position = {};
            if (data.position.x !== undefined) {
                this.position.x = setRangeValue(data.position.x);
            }
            if (data.position.y !== undefined) {
                this.position.y = setRangeValue(data.position.y);
            }
        }
        if (data.size !== undefined) {
            this.size.load(data.size);
        }
        loadProperty(this, "destroy", data.destroy);
        loadProperty(this, "orbits", data.orbits);
    }
}
