// @tsparticles/updater-destroy/browser/Options/Classes/Split.js
import { OptionsColor, deepExtend, executeOnSingleOrMultiple, isNull, loadProperty, } from "@tsparticles/engine";
import { SplitFactor } from "./SplitFactor.js";
import { SplitRate } from "./SplitRate.js";
export class Split {
    count = 1;
    factor = new SplitFactor();
    fillColor;
    fillColorOffset;
    particles;
    rate = new SplitRate();
    sizeOffset = true;
    strokeColor;
    strokeColorOffset;
    load(data) {
        if (isNull(data)) {
            return;
        }
        if (data.fillColor !== undefined) {
            this.fillColor = OptionsColor.create(this.fillColor, data.fillColor);
        }
        if (data.strokeColor !== undefined) {
            this.strokeColor = OptionsColor.create(this.strokeColor, data.strokeColor);
        }
        loadProperty(this, "count", data.count);
        this.factor.load(data.factor);
        this.rate.load(data.rate);
        this.particles = executeOnSingleOrMultiple(data.particles, particles => {
            return deepExtend({}, particles);
        });
        loadProperty(this, "sizeOffset", data.sizeOffset);
        if (data.fillColorOffset) {
            this.fillColorOffset = this.fillColorOffset ?? {};
            if (data.fillColorOffset.h !== undefined) {
                this.fillColorOffset.h = data.fillColorOffset.h;
            }
            if (data.fillColorOffset.s !== undefined) {
                this.fillColorOffset.s = data.fillColorOffset.s;
            }
            if (data.fillColorOffset.l !== undefined) {
                this.fillColorOffset.l = data.fillColorOffset.l;
            }
        }
        if (data.strokeColorOffset) {
            this.strokeColorOffset = this.strokeColorOffset ?? {};
            if (data.strokeColorOffset.h !== undefined) {
                this.strokeColorOffset.h = data.strokeColorOffset.h;
            }
            if (data.strokeColorOffset.s !== undefined) {
                this.strokeColorOffset.s = data.strokeColorOffset.s;
            }
            if (data.strokeColorOffset.l !== undefined) {
                this.strokeColorOffset.l = data.strokeColorOffset.l;
            }
        }
    }
}
