// @tsparticles/interaction-external-attract/esm/Options/Classes/Attract.js
import { isNull, loadProperty } from "@tsparticles/engine";
export class Attract {
    distance = 200;
    duration = 0.4;
    easing = "ease-out-quad";
    factor = 1;
    maxSpeed = 50;
    restore;
    speed = 1;
    constructor() {
        this.restore = {
            enable: false,
            delay: 0,
            speed: 0.08,
            follow: true,
        };
    }
    load(data) {
        if (isNull(data)) {
            return;
        }
        loadProperty(this, "distance", data.distance);
        loadProperty(this, "duration", data.duration);
        if (data.easing !== undefined) {
            this.easing = data.easing;
        }
        loadProperty(this, "factor", data.factor);
        if (data.maxSpeed !== undefined) {
            this.maxSpeed = data.maxSpeed;
        }
        loadProperty(this, "speed", data.speed);
        if (data.restore !== undefined) {
            this.restore.enable = data.restore.enable ?? this.restore.enable;
            this.restore.delay = data.restore.delay ?? this.restore.delay;
            this.restore.speed = data.restore.speed ?? this.restore.speed;
            this.restore.follow = data.restore.follow ?? this.restore.follow;
        }
    }
}
