// @tsparticles/animation-utils/tsparticles.animation.utils.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Utility v4.3.1 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.animation = global.__tsParticlesInternals.animation || {}, global.__tsParticlesInternals.animation.utils = global.__tsParticlesInternals.animation.utils || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    function checkDestroy(particle, destroyType, value, minValue, maxValue) {
        switch (destroyType) {
            case engine.DestroyType.max:
                if (value >= maxValue) {
                    particle.destroy();
                }
                break;
            case engine.DestroyType.min:
                if (value <= minValue) {
                    particle.destroy();
                }
                break;
        }
    }
    function initParticleNumericAnimationValue(options, pxRatio) {
        const valueRange = options.value, animationOptions = options.animation, res = {
            delayTime: engine.getRangeValue(animationOptions.delay) * engine.millisecondsToSeconds,
            enable: animationOptions.enable,
            value: engine.getRangeValue(options.value) * pxRatio,
            max: engine.getRangeMax(valueRange) * pxRatio,
            min: engine.getRangeMin(valueRange) * pxRatio,
            loops: 0,
            maxLoops: engine.getRangeValue(animationOptions.count),
            time: 0,
        }, decayOffset = 1;
        if (animationOptions.enable) {
            res.decay = decayOffset - engine.getRangeValue(animationOptions.decay);
            switch (animationOptions.mode) {
                case engine.AnimationMode.increase:
                    res.status = engine.AnimationStatus.increasing;
                    break;
                case engine.AnimationMode.decrease:
                    res.status = engine.AnimationStatus.decreasing;
                    break;
                case engine.AnimationMode.random:
                    res.status = engine.getRandom() >= engine.half ? engine.AnimationStatus.increasing : engine.AnimationStatus.decreasing;
                    break;
            }
            const autoStatus = animationOptions.mode === engine.AnimationMode.auto;
            switch (animationOptions.startValue) {
                case engine.StartValueType.min:
                    res.value = res.min;
                    if (autoStatus) {
                        res.status = engine.AnimationStatus.increasing;
                    }
                    break;
                case engine.StartValueType.max:
                    res.value = res.max;
                    if (autoStatus) {
                        res.status = engine.AnimationStatus.decreasing;
                    }
                    break;
                case engine.StartValueType.random:
                default:
                    res.value = engine.randomInRangeValue(res);
                    if (autoStatus) {
                        res.status = engine.getRandom() >= engine.half ? engine.AnimationStatus.increasing : engine.AnimationStatus.decreasing;
                    }
                    break;
            }
        }
        res.initialValue = res.value;
        return res;
    }
    function updateAnimation(particle, data, changeDirection, destroyType, delta) {
        const minLoops = 0, minDelay = 0, identity = 1, minVelocity = 0, minDecay = 1;
        if (particle.destroyed ||
            !data.enable ||
            ((data.maxLoops ?? minLoops) > minLoops && (data.loops ?? minLoops) > (data.maxLoops ?? minLoops))) {
            return;
        }
        const velocity = (data.velocity ?? minVelocity) * delta.factor, minValue = data.min, maxValue = data.max, decay = data.decay ?? minDecay;
        data.time ??= 0;
        const delayTime = data.delayTime ?? minDelay;
        if (delayTime > minDelay && data.time < delayTime) {
            data.time += delta.value;
            if (data.time < delayTime) {
                return;
            }
        }
        switch (data.status) {
            case engine.AnimationStatus.increasing:
                data.value += velocity;
                break;
            case engine.AnimationStatus.decreasing:
                data.value -= velocity;
                break;
        }
        if (data.velocity && decay !== identity) {
            data.velocity *= decay;
        }
        switch (data.status) {
            case engine.AnimationStatus.increasing:
                if (data.value >= maxValue) {
                    if (changeDirection) {
                        data.status = engine.AnimationStatus.decreasing;
                    }
                    else {
                        data.value -= maxValue;
                    }
                    data.loops ??= minLoops;
                    data.loops++;
                }
                break;
            case engine.AnimationStatus.decreasing:
                if (data.value <= minValue) {
                    if (changeDirection) {
                        data.status = engine.AnimationStatus.increasing;
                    }
                    else {
                        data.value += maxValue;
                    }
                    data.loops ??= minLoops;
                    data.loops++;
                }
                break;
        }
        checkDestroy(particle, destroyType, data.value, minValue, maxValue);
        if (!particle.destroyed) {
            data.value = engine.clamp(data.value, minValue, maxValue);
        }
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};

    exports.initParticleNumericAnimationValue = initParticleNumericAnimationValue;
    exports.updateAnimation = updateAnimation;

}));
delete (globalThis.window || globalThis).tsparticlesInternalExports;
